datumaro/components/media.py

```diff
@@ -217,49 +247,171 @@ def size(self) -> Optional[Tuple[int, int]]:
         super().set_crypter(crypter)
-        if isinstance(self._data, lazy_image):
-            self._data._crypter = crypter
+        if isinstance(self.__data, lazy_image):
+            self.__data._crypter = crypter
+
+
+class ImageFromData(FromDataMixin, Image):
+    def save(
+        self,
+        fp: Union[str, io.IOBase],
+        ext: Optional[str] = None,
+        crypter: Crypter = NULL_CRYPTER,
+    ):
+        data = self.data
+        if data is None:
+            raise ValueError(f"{self.__class__.__name__} is empty.")
+        new_ext = self._get_ext_to_save(fp, ext)
+        if isinstance(fp, str):
+            os.makedirs(osp.dirname(fp), exist_ok=True)
+        save_image(fp, data, ext=new_ext, crypter=crypter)
+
+
+class ImageFromNumpy(ImageFromData):
+    def __init__(
+        self,
+        data: Union[Callable[[], bytes], bytes],
+        *args,
+        **kwargs,
+    ):
+        super().__init__(data=data, *args, **kwargs)
 
     @property
-    def path(self) -> str:
-        """Path to the media file"""
-        return self._path
+    def data(self) -> Optional[np.ndarray]:
+        """Image data in BGR HWC [0; 255] (float) format"""
+
+        data = super().data
+
+        if isinstance(data, np.ndarray):
+            data = data.astype(np.float32)
+        if self._size is None and data is not None:
+            if not 2 <= data.ndim <= 3:
+                raise MediaShapeError("An image should have 2 (gray) or 3 (rgb) dims.")
+            self._size = tuple(map(int, data.shape[:2]))
+        return data
+
+    @property
+    def has_size(self) -> bool:
+        """Indicates that size info is cached and won't require image loading"""
+        return self._size is not None or isinstance(self._data, np.ndarray)
+
+
+class ImageFromBytes(ImageFromData):
+    _FORMAT_MAGICS = (
+        (b"\x89PNG\r\n\x1a\n", ".png"),
+        (b"\xff\xd8\xff", ".jpg"),
+        (b"BM", ".bmp"),
+    )
+
+    def __init__(
+        self,
+        data: Union[Callable[[], bytes], bytes],
+        *args,
+        **kwargs,
+    ):
+        super().__init__(data=data, *args, **kwargs)
+
+        if self._ext is None and isinstance(data, bytes):
+            self._ext = self._guess_ext(data)
+
+    @classmethod
+    def _guess_ext(cls, data: bytes) -> Optional[str]:
+        return next(
+            (ext for magic, ext in cls._FORMAT_MAGICS if data.startswith(magic)),
+            None,
+        )
+
+    @property
+    def data(self) -> Optional[np.ndarray]:
+        """Image data in BGR HWC [0; 255] (float) format"""
 
+        data = super().data
 
-class ByteImage(Image):
+        if isinstance(data, bytes):
+            data = decode_image(data)
+        if isinstance(data, np.ndarray):
+            data = data.astype(np.float32)
+        if self._size is None and data is not None:
+            if not 2 <= data.ndim <= 3:
+                raise MediaShapeError("An image should have 2 (gray) or 3 (rgb) dims.")
```

Comment:
Here too.

Reply:
It was fixed in https://github.com/openvinotoolkit/datumaro/pull/921/commits/e19ad25b3abdf7c8103198a671e502e1f20d71d3